R1cs bulletproofs (#236)
tari-labs/tari-university: src/cryptography/r1cs-bulletproofs/mainreport.md

> 

Reviewer: We are still missing some expected content as per http://localhost:3000/preface/style-guide.html#proposed-layout, i.e. **Conclusions, Observations, Recommendations** and **Contributors **.

> # **Rank-1 Constraint System with Application to Bulletproofs**

Reviewer: This TOC should link all headings in the report; not all are represented here

> - [Introduction](introduction) 

Reviewer: These links are not working, see example use in other reports

According to Amir Shpilka and Amir Yehudayoff [[3.]], arithmetic circuits are a highly structured model of computation compared to Boolean circuits. For instance, when studying arithmetic circuits one is interested in _syntactic_ computation of polynomials, whereas in studying Boolean circuits one is interested in the _semantics_ of the computation. In other words, in the Boolean case one is not interested in any specific _polynomial representation_ of the function but rather one just wants to compute some representation of it, while in the arithmetic world one focuses on a _specific representation_ of the function. 

For our purposes, arithmetic circuits have as inputs variables  $\large{ x_1, ... , x_n }$, and computations are performed using the arithmetic operations  $ + $  and  $ × $, and may involve constants from a field  $\mathcal{F}_p$. Complexity measures associated with such circuits are _size_ and _depth_, which capture _the number of gates_ in a circuit and _the maximal distance between an input and an output_, respectively [[3.]].

> ### Definition 1 (Arithmetic circuits) [[3.]]

Reviewer: Please remove references from headings and add it to the paragraph text.

A typical _computational problem_ would involve finding the solution to, let's say,  $\large{x^2_1 x_2 + x_1 + 1 = 22}$.  Or, in a proof of knowledge scenario, the prover has to prove to the verifier that he has the correct solution to such an equation. 

Following the wires, in Example 1.2, shows that an arithmetic circuit actually _breaks down_ the given computation into _smaller equations_ corresponding to each _gate_; 

​				$\large{ u = x_1*x_1} $ ,   $ \large{ v = u*x_2 }$ ,   $ \large{ y = x_1 + 1 } $ , and  $ \large{ z = v + y } $ . 

The variables  $ \large{ u , v }$  and $ \large{ y }$  are called _auxiliary variables_ or _low level variables_, while  $\large{ z }$  is the _output_ of  $ \mathcal{A} $.  Thus, in addition to computing polynomials naturally, an arithmetic circuit helps in reducing a computation to a _low level_ language involving only _two_ variables, _one operation_, and an _output_. 

Although "Arithmetic circuits are a highly structured model of computation compared to Boolean circuits," according to Amir Shpilka and Amir Yehudayoff [[3.]], "we do not know how to efficiently reconstruct a circuit using only queries to the polynomial it computes." 

> Zero-knowledge proofs, such as zkSNARKs and Bulletproofs, require _statements to be proved_ be expressed in their simplest terms for efficiency. Alex Pinto [[5.]] mentions that "the ZK SNARK end-to-end journey is to create a _function_ to write proofs about," but they "must work with specific constructs." That is why arithmetic circuits are so important in making zero-knowledge more efficient, "these _functions_ have to be specified as sequences of very simple terms, namely, additions and multiplications of only two terms in a particular field" [[5.]].

Reviewer: Limit line lengths to 120 characters, except in tables and within links. Apply throughout.

Author: I'm not sure I understand what problem we are solving here. Do I just insert "<\br>" after every 120 characters? everywhere?

Reviewer: Use a text editor to insert newlines. Markdown recognizes consecutive lines to be in the same paragraph, while paragraphs are separated by an empty line. See example below:

```
Zero-knowledge proofs, such as zkSNARKs and Bulletproofs, require _statements to be proved_ be expressed in their 
simplest terms for efficiency. Alex Pinto [[5.]] mentions that "the ZK SNARK end-to-end journey is to create a 
_function_ to write proofs about," but they "must work with specific constructs." That is why arithmetic circuits are so 
important in making zero-knowledge more efficient, "these _functions_ have to be specified as sequences of very simple 
terms, namely, additions and multiplications of only two terms in a particular field" [[5.]].
```

Reviewer: But just check, maybe @anselld already did this when she performed her technical editing. Then this comment is for future use.

The reason we limit line lengths is to enable more efficient and pinpoint reviews for text files like Markdown in GitHub .

Reviewer 2: Yes, I did limit the line lengths, @hansieodendaal and @Empiech007 .

| 6.   | The zero-knowledge proof involves _conversion_ of the Arithmetic circuit into a Rank-1 Constraint system. | The mathematical expression of a _Hadamard relation_ is closely related to an _Arithmetic circuit_. The use of this relation plus linear constraints as a single constraint amounts to using a constraint system. | Although _Arithmetic circuits_ are not explicitly used here, the Hadamard relation remains the same as first seen in Bulletproofs, moreso in the _inner-product proof_. |
| 7.   | The zero-knowledge proof here is for NP statements based on the _difficulty_ of _the discrete logarithm problem_. | As mentioned above, Bulletproofs extend zero-knowledge proofs (such as _range proofs_) to proofs on arbitrary statements. They are also described as _short non-Interactive proofs for Arithmetic Circuits without a Trusted Setup_. | Interstellar is building an API that allows developers to choose their own collection of gadgets suitable for the protocol they wish to develop. |

> ​				**Table 2 : Comparison of three Research Works Zero Knowledge Proofs**   

Reviewer: - Table captions go in front of the table
- This seems to be the 1st table not the 2nd

> ## Constraint Systems for ZK SNARKs 

Reviewer: Suggest not to use the 'curated content' style here. The overview and summary should be sufficient. Just make sure appropriate references are added.

> <img src="sources/alex-pinto.png" alt="alex-pinto" style="zoom: 15%;" /> 

Reviewer: ```suggestion
```

> **Alex Pinto**

Reviewer: ```suggestion
```

> Developer and researcher in the blockchain space

Reviewer: ```suggestion
```

> #### Overview 

Reviewer: ```suggestion
```

Now, how exactly does one go _**from an Arithmetic circuit to an R1CS**_ ? 

Eli Ben-Sasson et al. mentioned that the difference in computational power among the parties makes _succinct verification_ a requirement [[8.]]. Hence the pursuit of short and fast proofs systems. These make it possible for a verifier to check a non-deterministic polynomial-time computation in a much shorter time than the time required to run the computation when given a valid NP witness. They also state that the difficulty in studying the efficiency of proofs for NP statements is the _problem of representation_. Arithmetic circuits and the R1CS are some of such _representations_ of statements to be proved. 

> #### Summary 

Reviewer: ```suggestion
```

| 5.   | Prover creates the proof.                                    | `let proof = prover.prove().unwrap();`                       |
| 6.   | Instantiation of the Verifier using the Pedersen<br/> commitments and Bulletproof generators, <br/>and creates its own transcript. | `let mut verifier_transcript = Transcript::new(b"Factors");`<br/>`let mut verifier = Verifier::new(&bp_gens, &pc_gens, &mut verifier_transcript);` |
| 7.   | Verifier _records_ commitments<br/>for p and q sent by prover in<br/>the transcript, and creates variables for them similar to the prover's. | `let var_p = verifier.commit(commitments.0);`<br/>`let var_q = verifier.commit(commitments.1);` |
| 8.   | Verifier _constrains_ variables corresponding to the<br/>commitments. | `let (_, _, o) =  verifier.multiply(var_p.into(), var_q.into());`<br/>`let r_lc: LinearCombination = vec![(Variable::One(), r.into())].iter().collect();`<br/>`verifier.constrain(o -  r_lc);` |
| 9.   | Verifier _verifies_ the _proof_.                             | `verifier.verify(&proof)`                                    |

​			**Table 3 : Bulletproof Constraint Example**

> ## Zero-Knowledge Proofs: Privacy & Scaling

Reviewer: As per previous:

Suggest not to use the _'curated content'_ style here. The overview and summary should be sufficient. Just make sure appropriate references are added.

For further information and a more inclusive list of references on zero-knowledge proofs (with links) see [[18.]]." It contains a tabular list of "Implementations of proving systems" which includes [bellman](https://github.com/zkcrypto/bellman), [dalek bulletproofs](https://github.com/dalek-cryptography/bulletproofs) and [adjoint-io bulletproofs](https://github.com/adjoint-io/bulletproofs). In line with this report, one also finds a list of "Libraries for writing circuits (or gadgets)" such as [libsnark](https://github.com/scipr-lab/libsnark),  [ZoKrates](https://github.com/Zokrates/ZoKrates), [snarky](https://github.com/o1-labs/snarky), and [ZkMV](https://github.com/stellar/slingshot/tree/main/zkvm). 

There are standardization efforts made by the community with [the third _ZKProof workshop_ coming up in April 4 - 6, 2020 in London](https://zkproof.org/), and communication made with NIST in this regard as seen in the letter [here](https://zkp.science/docs/Letter-to-NIST-20160613-Advanced-Crypto.pdf).  

> ## References 

Reviewer: Please fix all references as per https://raw.githubusercontent.com/tari-labs/tari-university/master/src/cryptography/bulletproofs-and-mimblewimble/MainReport.md

> [[1.](https://medium.com/@cathieyun/building-on-bulletproofs-2faa58af0ba8)] Cathie Yun, "Building on Bulletproofs," https://medium.com/@cathieyun/building-on-bulletproofs-2faa58af0ba8 Date accessed: 2019‑01‑03 

Reviewer: ```suggestion
[[1]] Cathie Yun, "Building on Bulletproofs" [online]. Available: <https://medium.com/@cathieyun/building-on-bulletproofs-2faa58af0ba8> Date accessed: 2019&#8209;01&#8209;03 .
```

Reviewer: Note: We replace `-` in the date with `&#8209;`, which is non-breaking accross lines

> [1.]: https://medium.com/@cathieyun/building-on-bulletproofs-2faa58af0ba8

Reviewer: ```suggestion
[1]: https://medium.com/@cathieyun/building-on-bulletproofs-2faa58af0ba8 "Building on Bulletproofs"
```

"Building on Bulletproofs"

> [[2.]]  Alan Szepieniec and Bart Preneel, "Generic Zero-Knowledge and Multivariate Quadratic Systems," 2016. https://pdfs.semanticscholar.org/06c8/ea507b2c4aaf7b421bd0c93e6145e3ff7517.pdf?_ga=2.124585865.240482160.1578465071-151955209.1571053591 Date accessed: 2020‑01‑08. 

Reviewer: ```suggestion
[[2]]  Alan Szepieniec and Bart Preneel, "Generic Zero-Knowledge and Multivariate Quadratic Systems" [online]. Available: <https://pdfs.semanticscholar.org/06c8/ea507b2c4aaf7b421bd0c93e6145e3ff7517.pdf?_ga=2.124585865.240482160.1578465071-151955209.1571053591> Date accessed: 2020&#8209;01&#8209;08. 
```

> [2.]: https://pdfs.semanticscholar.org/06c8/ea507b2c4aaf7b421bd0c93e6145e3ff7517.pdf?_ga=2.124585865.240482160.1578465071-151955209.1571053591

Reviewer: ```suggestion
[2]: https://pdfs.semanticscholar.org/06c8/ea507b2c4aaf7b421bd0c93e6145e3ff7517.pdf?_ga=2.124585865.240482160.1578465071-151955209.1571053591
"Generic Zero-Knowledge and 
Multivariate Quadratic Systems"
```

##### 

> 

Author: Done

  - [R1CS Definition for zkSNARKs](#r1cs-definition-for-zksnarks)
  - [Example of Rank-1 Constraint System](#example-of-rank-1-constraint-system) 
- [From Arithmetic Circuits to Programmable Constraint Systems for Bulletproofs ](#from-arithmetic-circuits-to-programmable-constraint-systems-for-bulletproofs) 
- [Interstellar's Bulletproof Constraint System](#interstellars-bulletproof-constraint-system)
  - [Overview](#overview-2) 
  - [Easy to Build Constraint Systems](#easy-to-build-constraint-systems) 
  - [About Gadgets](#about-gadgets) 
  - [Interstellar's Concluding Remarks](#interstellars-concluding-remarks) 
  - [R1CS Factorization Example for Bulletproofs ](#r1cs-factorization-example-for-bulletproofs)
- [Conclusions, Observations and Recommendations](#conclusions-observations-and-recommendations)
- [References](#references) 
- [Contributors](#contributors) 

## Introduction 

This report explains the technical underpinnings of Rank-1 Constraint Systems (R1CSs) as applied 
to Bulletproofs. 

The literature on the use of R1CSs in zero-knowledge (ZK) proofs, for example in zk-SNARKs, shows that this mathematical tool is used simply as one part of many in 

> a complex process towards achieving the proof. Not much attention is given to it, not even in explaining 

Author: Done

what "rank-1" actually means. Although the terminology is similar to the traditional _rank of a matrix_ in 
linear algebra, examples on the Internet do not yield a _reduced matrix with only one non-zero row or column_. 

R1CSs became more prominent for Bulletproofs due to research work done by Cathie Yun and her colleagues at Interstellar. The constraint 
systems, in particular R1CSs, are used as an add-on to Bulletproof protocols. The title of 
Yun's article, "Building on Bulletproofs" 
[[1]] suggests this is true. One of the Interstellar team's goals is to use the constraint system in their 
Confidential Asset Protocol called the Cloak and in their envisaged Spacesuit. 
Despite their work on using R1CS being research in progress, their detailed notes on constraint systems and their 
implementation in RUST are available in [[2]]. 

Constructions of zero-knowledge proofs such as Bulletproofs involve arithmetic circuits and R1CSs, 

> either explicitly or implicitly. 

Author: This is a remnant of the zkSNARK perspective I initially had in the report. Sentence removed.

The aim of this report is to: 

- highlight the connection between arithmetic circuits and R1CSs; 
- clarify the difference R1CSs make in Bulletproofs and in range proofs; 
- compare ZK Proofs for arithmetic circuits and Programmable Constraint System; 

> - explain what is meant by the following in [[3]] - "We note that a range proof 

Author: I am not sure how to resolve this one? Should I not call it the aim, perhaps?

> - explain what is meant by the following in [[3]] - "We note that a range proof 

Author: The report scope is itself narrow, so it's hard not to be specific.

> - explain what is meant by the following in [[3]] - "We note that a range proof 

Author: I do not agree with the 'abstract' part. 
The language used makes the aims as lucid as it can get; "highlight the connection", "clarify the difference", "compare", and "explain what is meant by the following"

> - explain what is meant by the following in [[3]] - "We note that a range proof 

Reviewer: I think the 4th bullet should just be removed. The first three bullets set a nice scope, whereas the 4th bullet mentions detail that most readers will just think about "so what".

> - explain what is meant by the following in [[3]] - "We note that a range proof 

Author: A reader who is interested in the research developments might be curious to know what improvements have been made by Bunz et al to the Bootle et al's paper, and how Bunz et al achieve 'the elimination' of the 'commitment opening algorithm'. 

There are issues with this particular change as it seems biased towards the prover (Cathie Yun gave a caveat in [[11]](https://medium.com/interstellar/programmable-constraint-systems-for-bulletproofs-365b9feb92f7)), in the paragraph just above "Gadgets". This is related to the 1st bullet under "Interstellar's Concluding Remarks".)

That being said, I suggest appending the following sentence to the 4th bullet: "and how Bunz et al achieve this elimination".

> - explain what is meant by the following in [[3]] - "We note that a range proof 

Reviewer: Although this is interesting, and all that you said should be reflected in the section "Definition of Constraint System", for it to be an aim of the report it cannot be covered in a single paragraph. Furthermore, the specific nature of the statement cannot make it an aim of the report. An aim could have been - "Discusses improvements made by Bunz et al [4] to the work done by Bootle et al [5]" - but this is definitely not the aim of the report.

  using the protocol of Bootle et al. [[4]] would have required implementing the commitment opening 
  algorithm as part of the verification circuit, which we are able to eliminate." 

## Arithmetic Circuits

> ### Overview 

Author: Cool. That will work.

can be represented as an arithmetic circuit over any finite field $ \mathcal{F}_p$ , [[5]]. 

According to [[6]], arithmetic circuits are a highly structured model of computation 
where one is interested in _syntactic_ 
computation of polynomials. That is, focusing on specific representation of the function being computed in order to improve efficiency. 

### Definition of Arithmetic Circuit 

An **arithmetic circuit** $\mathcal{A}$ over the field $\mathcal{F}$ and the set of variables 
$X = \lbrace  {x_1,\dots,x_n} \rbrace$ is a directed acyclic graph such that the vertices of $\mathcal{A}$ are called 
_gates_, while the edges are called _wires_ [[6]]: 

- A gate is said to be of in-degree  $l$  if it has  $l$  incoming wires, and similarly, of out-degree  $k$  if 
  it has  $k$  outgoing wires. 
- Every gate in  $\mathcal{A}$  of in-degree 0 is labeled by either a variable  ${x_i}$  or some field 
  element from  $\mathcal{F}$.  Such a gate is called an  input gate. Every gate of out-degree  0  is called an 
  output gate or the root. 
- Every other gate in  $\mathcal{A}$  is labeled by either  $\otimes$  or  $\oplus$ , and called a 

>   multiplication gate or addition gate, respectively. We are interested in binary circuits, i.e. circuits where 

Author: Omitted the term "binary circuits" and used a descriptive explanation.

  each gate has two input wires and one output wire. 
- An arithmetic circuit is called a formula if it is a directed tree whose edges are directed from the leaves 
  to the root. 
- The depth of a node is the number of edges in the longest directed path between the node and the root.
- The size of  $\mathcal{A}$ , denoted  $|\mathcal{A}|$ , is the number of wires, i.e. edges, in  $\mathcal{A}$. 

> For our purposes, arithmetic circuits have as inputs variables $ x_1, \dots , x_n $. Computations are 

Author: Paragraph removed.

performed using the arithmetic operations $ + $ and $ × $, and may involve constants from some field $\mathcal{F}$. 
Complexity measures associated with such circuits are size and depth, which capture the number of gates in a 
circuit and the maximum distance (the number of edges in the longest directed path) between an input and an output, respectively [[6]].

A typical multiplication gate has a left input  $a_L$ , a right input  $a_R$  and 

> an output  $a_O$. Also, we note that  $ a_L \cdot a_R - a_O = 0 $.

Reviewer: ```suggestion
an output  $a_O$ (Figure&nbsp;1). Also, we note that  $ a_L \cdot a_R - a_O = 0 $.
```

<div align="center"><b>
<img src="sources/basic-multiplication-gate.png" alt="basic-multiplication-gate" style="zoom:67%;" />
</b></div> 
<div align="center"><b>Figure 1: Typical Multiplication Gate</b></div> 

> We note that in cases where the inputs and outputs are all vectors of  

Author: Not sure what this commit changes. I have rewritten the sentence without pronouns.

> We note that in cases where the inputs and outputs are all vectors of  

Reviewer: The suggested change here was a space after the HTML

$n$  components, i.e.  $\mathbf{a_L} = ( a_{L, 1}, a_{L, 2} , \dots , a_{L, n})$, 
$\mathbf{a_R} = ( a_{R, 1}, a_{R, 2} , \dots , a_{R, n})$ 
and  $\mathbf{a_O} = ( a_{O, 1}, a_{O, 2} , \dots , a_{O, n})$, 
then multiplication of $\mathbf{a_L}$ and $\mathbf{a_R}$ is defined as an entry-wise 
product called the **Hadamard product**; 
$$
\mathbf{a_L}\circ \mathbf{a_R} = \big(( a_{L, 1} \cdot a_{R, 1} ) , ( a_{L, 2} \cdot a_{R, 2} ) , \dots , ( a_{L, n} \cdot 
a_{R, n} ) \big) =  \mathbf{a_O}
$$

The equation ${ \mathbf{a_L}\circ \mathbf{a_R} = \mathbf{a_O} }$ is referred to as a multiplicative constraint, 
but is also known as the Hadamard relation [[4]]. 

### Example of Arithmetic Circuit 

> An arithmetic circuit computes a polynomial in a natural way, as shown in this example. Consider the following 

Reviewer: ```suggestion
An arithmetic circuit computes a polynomial in a natural way, as shown in this example. Consider the 
```

> arithmetic circuit  $\mathcal{A}$  with inputs  $\lbrace x_1 , x_2 , 1 \rbrace$  over some field  $\mathcal{F}$ : 

Reviewer: ```suggestion
arithmetic circuit  $\mathcal{A}$  with inputs  $\lbrace x_1 , x_2 , 1 \rbrace$  over some field  $\mathcal{F}$ (Figure&nbsp;2): 
```

<div align="center"><b>
<img src="sources/polynomial-eg-ac.png" alt="polynomial-eg-ac" style="zoom:67%;" />
</b></div> 
<div align="center"><b>Figure 2: Arithmetic Circuit</b></div> 

> The output of $\mathcal{A}$ above is the polynomial $x^2_1 \cdot x\_2 + x\_1 + 1 $ of total degree three. 

Reviewer: ```suggestion

The output of $\mathcal{A}$ above is the polynomial $x^2_1 \cdot x\_2 + x\_1 + 1 $ of total degree three. 
```

Reviewer: Better to have an empty line after HTML codes

Author: Done

A typical computational problem would involve finding the solution to, let's say,  $x^2\_1 \cdot x\_2 + x\_1 + 1 = 22$.  
Or, in a proof of knowledge scenario, the prover has to prove to the verifier that they have the correct solution to such 
an equation. 

Following the wires in the example shows that an arithmetic circuit actually breaks down the given computation into 
smaller equations corresponding to each gate: 
$$
u = x_1 \cdot x_1 \quad \text{,} \quad v = u \cdot x_2 \quad \text{,} \quad y = x_1 + 1 \quad \text{and} \quad z = v + y
$$

The variables $u, v$ and $ y $ are called auxiliary variables or low-level variables, while 
${ z }$  is the output of  $ \mathcal{A} $ . Thus, in addition to computing polynomials naturally, an 
arithmetic circuit helps in reducing a computation to a low-level language involving only two variables, 
one operation and an output. 

> Zero-knowledge proofs in general require that statements to be proved are expressed in their 

Author: fixed

> Zero-knowledge proofs in general require that statements to be proved are expressed in their 

Author: Fixed

### Overview

Although a computational problem is typically expressed in terms of a high-level programming language, a ZK proof requires 
it to be expressed in terms of a set of quadratic constraints, which are closely related to circuits of logical gates. 

Other than in Bulletproof constructions, R1CS types of constraint systems have been featured in several 
constructions of zk-SNARKs. At times they were simply referred to as quadratic constraints or quadratic equations; refer to 
[[5]], [[8]] and [[9]]. 

### Definition of Constraint System 

A constraint system was originally defined by Bootle 
et al. in [[4]]. The Dalek team give a more general definition of a constraint system in [[2]]: 

"A **constraint system** is a collection of arithmetic constraints over a set of variables. There are two kinds of 
variables in the constraint system:

1. ${m}$  **high-level** variables, the input secrets  ${ \mathbf{v}}$, 
2. $ n$  **low-level** variables, the internal input vectors  ${ \mathbf{a}_L}$  and  
   ${ \mathbf{a}_R},$  and output vectors  ${ \mathbf{a}_O } $  of the multiplication gates." 

Specifically, an R1CS is a system that consists of two sets of constraints: 

> - ${ n}$  multiplicative constraints,  $ \mathbf{ a_L \circ a_R = a_O } $,  where  "${ \circ }$"  is the Hadamard product; and

Author: Removed

Note that it was Bootle et al. who first expressed arithmetic circuit satisfiability in terms of the Hadamard relation 
and linear constraints [[2]]. In their definition, the above linear constraints are written as:
$$
\mathbf{W_L\cdot { a_L} + W_R\cdot { a_R} + W_O\cdot { a_O } = c }
$$

That is, without the vector  $\mathbf{v}$  and its weight  $\mathbf{W_V} $. 

As to why there is a vector  $\mathbf{v}$  and its weight  $\mathbf{W_V}$  in the definition, Bunz et al. 
explain that "we include additional commitments  $V_i$  as part of our statement, and give a protocol for a more general 
relation, where the linear consistency constraints include the openings  ${ v_j}$  of the commitments  $V_j$", 
refer to page 24 of [[3]]. Their definition of a constraint system incorporates a secret vector  $\mathbf{v}$  and its weight  $\mathbf{W_V}$,  because commitments  $V_i$  of components  ${ v_i}$  of  $\mathbf{v} = {(v_1, v_2, \dots , v_m )}$  are included among the inputs. We note that 
Bulletproofs use the Pedersen commitment scheme.

> ### R1CS Definition for zkSNARKs 

Author: Done

> This paragraph provides a simplified definition of an R1CS as it applies to zk-SNARKs [[10]]. 

Author: Done

$$
where
$$\langle {\mathbf{a_L} , \mathbf{s}} \rangle = a_{L, 1} \cdot s_1 + a_{L, 2} \cdot s_2 + \cdots + a_{L, n} \cdot s_n $$
which is the innerproduct of the vectors 
$ \mathbf{a\_{L}}  =  (a\_{L,1}, a\_{L,2}, ... , a\_{L,n} )$
and
$ {\mathbf{s}} = (s\_1, s\_2, ... , s\_n) $.

### Example of Rank-1 Constraint System 

One solution to the equation ${x^2_1 x_2 + x_1 + 1 = 22}$, from the preceding example of an arithmetic circuit, is ${ x_1 = 3}$ 
and ${ { x_2 = 2 }}$ belonging to the appropriate field ${ \mathcal{F}}$. Thus the solution vector 
${ { s = ( const , x_1 , x_2 , z , u , v , y )}}$ becomes ${ { s = ( 1 , 3 , 2 , 22 , 9 , 18 , 4 )}}$. 

It is easy to check that the R1CS for the computation problem in the preceding example is as follows (one need only test 
if ${ \langle {\mathbf{a_L} , \mathbf{s}} \rangle  \cdot  \langle {\mathbf{a_R} , \mathbf{s}} \rangle - \langle {\mathbf{a_O} , \mathbf{s}} \rangle = 0}$ 
for each equation). 

<div align="center"><b>Table 1: Equations and Rank-1 Constraint System Vectors</b></div> 

> | Equation                        | Rank-1 Constraint System Vectors                             |

Author: Spot on. 
Fixed.

| ------------------------------- | ------------------------------------------------------------ |
| ${ u = x_1\cdot x_1}$          | $ {\bf{a_L}} = ( 0 , 1 , 0 , 0 , 0 , 0 , 0 ) , \ \ {\bf{a_R}} = ( 0 , 1 , 0 , 0 , 0 , 0 , 0  ) ,\ \ {\bf{a_O}} = ( 0 , 0 , 0 , 0 , 1 , 0 , 0  ) $ |
| $ { v = u\cdot x_2 }$          | $ {\bf{a_L}} = ( 0 , 0 , 0 , 0 , 1 , 0 , 0 ) ,\ \ {\bf{a_R}} = ( 0 , 1 , 0 , 0 , 0 , 0 , 0  ),\ \ {\bf{a_O}} = ( 0 , 0 , 0 , 0 , 0 , 1 , 0 )  $ |
| $ { y = 1\cdot( x_1 + 1 ) } $ | ${\bf{a_L}} = ( 1 , 1 , 0 , 0 , 0 , 0 , 0 ),\ \ {\bf{a_R}} = ( 1 , 0 , 0 , 0 , 0 , 0 , 0 ),\ \ {\bf{a_O}} = ( 0 , 0 , 0 , 0 , 0 , 0 , 1 ) $ |
| $ { z = 1\cdot( v + y )} $    | ${\bf{a_L}} = ( 0 , 0 , 0 , 0 , 0 , 1 , 1 ),\ \ {\bf{a_R}} = ( 1 , 0 , 0 , 0 , 0 , 0 , 0 ),\ \ {\bf{a_O}} = ( 0 , 0 , 0 , 1 , 0 , 0 , 0 )$ |

In a more formal definition, an **R1CS** is a set of three matrices  ${\bf{ A_L , A_R }}$  and  ${\bf A_O}$, 
where the rows of each matrix are formed by the corresponding vectors  $ {\bf{a_L }}$,  ${ \bf{a_R }}$  and 
${ \bf{a_O}} $, respectively, as shown in Table 1: 

> $$

Author: Okay

| 1.   | Introduces the Hadamard relation and linear constraints.     | Turns the Hadamard relation and linear constraints into a single linear constraint, and these are in fact the R1CS. | Generalizes constraint systems and uses what is called gadgets as building blocks for constraint systems. |
| 2.   | Improves on Groth's work [[12]] on ZK proofs. Reducing a $\sqrt{N}$  complexity to  $6log_2(N) + 13$, where $N$  is the circuit size. | Improves on Bootle et al.'s work [[4]]. Reducing a $2log_2(N) + 13$ complexity to  $6log_2(N) + 13$, where $N$ is the circuit size. | Adds constraint systems to Bunz et al.'s work on Bulletproofs, which are short proofs, and the complexity advantage is seen in proving several statements at once. |
| 3.   | Introduces logarithm-sized inner-product ZK proofs.          | Introduces Bulletproofs, extending proofs to proofs of arbitrary statements. The halving method is used on the inner-products, resulting in the above reduction in complexity. | Introduces gadgets that are actually add-ons to an ordinary ZK proof. A range proof is an example of a gadget. |

> | 4.   | Uses Fiat-Shamir heuristics in order to achieve non-interactive ZK proofs. | Bulletproofs have no trusted setup, as they use  the Fiat Shamir heuristics to achieve non-interaction. | Merlin transcripts are specifically used for a Fiat-Shamir transformation to achieve non-interaction. |

Author: Sorted.

article, "Programmable Constraint Systems for Bulletproofs"
[[11]]. However, it was Bootle et al. [[4]] who first used the Hadamard relation and linear constraints, which together 
form the constraint system as formalized by the Interstellar team. Most of the mathematical background of these 
constraints and bulletproofs is contained in Bunz et al.'s paper [[3]].

Dalek's constraint system, as defined earlier in [Definition of Constraint System](#definition-of-constraint-system), is a collection of arithmetic constraints of two types, 
multiplicative constraints and linear constraints, over a set of high-level and low-level variables. 

 

### Easy-to-build Constraint Systems 

In this bulletproofs framework, a prover can build a constraint system in two steps:

- Firstly, by committing to secret 
  inputs and allocating high-level variables corresponding to the inputs.
- Secondly, by selecting a suitable combination 
  of multiplicative constraints and linear constraints, as well as requesting a random scalar in response to the 
  high-level variables already committed [[2]]. 

> Reference 

Reviewer: ```suggestion
```

Author: Not sure what to change here.

> [[N16]] gives an excellent outline of ZK proofs that use Bulletproofs: 

Reviewer: ```suggestion
Reference [[16]] gives an excellent outline of ZK proofs that use Bulletproofs: 
```

1. The prover commits to a value(s) that they want to prove knowledge of. 
2. The prover generates the proof by enforcing the constraints over the committed values and any additional 
   public values. The constraints might require the prover to commit to some additional variables. 
3. The Prover sends the verifier all the commitments made in step 1 and step 2 along with the proof from step 2. 
4. The verifier now verifies the proof by enforcing the same constraints over the commitments plus any public 
   values.

### About Gadgets 

Consider a verifiable shuffle: given two lists of committed values  ${ x_1, x_2, . . . , x_n}$  and 

> ${ y_1, y_2, . . . , y_n} ,$ prove that the second list is a permutation of the first. Bunz et al. (page 5 of [[3]]) mention that 

Reviewer: ```suggestion
${ y_1, y_2, . . . , y_n} ,$ prove that the second list is a permutation of the first. Bunz et al. ([[3], page 5]) mention that 
```

Reviewer: Use of references as per https://tlu.tarilabs.com/preface/style-guide.html#referencing-standard and https://guides.lib.monash.edu/citing-referencing/ieee

the use of bulletproofs improves the complexity of such a verifiable shuffle to size $\mathcal{O}(log(n))$ compared to 
previous implementation results. Although not referred to as a _gadget_ in the paper, this is in 
fact a shuffle gadget. The term gadget was used and popularized by the Interstellar team, who introduced gadgets 
as building blocks of constraint systems; refer to [[1]]. 

> A _shuffle gadget_ is any function whose outputs are but a permutation of its inputs. By definition of a permutation, the 

Reviewer: ```suggestion
A _shuffle gadget_ (Figure&nbsp;3) is any function whose outputs are but a permutation of its inputs. By definition of a permutation, the 
```

number of inputs to a shuffle gadget is always the same as the number of outputs. 

<div align="center"><b> 
<img src="sources/simple-shuffle-diagram.png" style="zoom:67%;" /> 
</b></div> 
​				

> <div align="center"><b>Figure 3: Simple Shuffle Gadgets with Two Inputs [[1]]</b></div> 

Reviewer: ```suggestion
<div align="center"><b>Figure 3: Simple Shuffle Gadgets with Two Inputs [<a href="https://medium.com/@cathieyun/building-on-bulletproofs-2faa58af0ba8" 
title="Building on Bulletproofs">1</a>]</b></div> 
```

Reviewer: In HTML code, a Markdown link like `[[1]]` must be inserted as
[<a href="the_URL_link" title="the_title">1</a>]

> The Interstellar team mentions other gadgets: “merge”, “split” and a “range proof”, that are implemented in their 

Reviewer: ```suggestion

The Interstellar team mentions other gadgets: “merge”, “split” and a “range proof”, that are implemented in their 
```

Author: Got it.

> Confidential Assets scheme called the _Cloak_. Just as a shuffle gadget creates constraints which prove that 

Reviewer: `_Cloak_` does not render properly

Reviewer: An empty line is required between this paragraph and the previous HTML codes.

Author: Sorted

 choose their own collection of gadgets suitable for the protocol they wish to develop. 

### Interstellar's Concluding Remarks 

Cathie Yun reports in [[11]] that their "work on Cloak and Spacesuit is far from complete" and mentions that they 
still have two more goals to achieve: 

- Firstly, in order to "ensure that challenge-based variables cannot be inspected" and prevent the user from 
  accidentally breaking soundness of their gadgets, the Bulletproofs protocol needs to be slightly extended, enabling it 
  to commit "to a portion of low-level variables using a single vector Pedersen commitment without an overhead of 
  additional individual high-level Pedersen commitments" [[11]]. 
- Secondly, to "improve privacy in Cloak" by enabling "multi-party proving of a single constraint system". That is, 
  "building a joint proof of a single constraint system by multiple parties, without sharing their secret inputs with 
  each other".

All-in-all, Yun regards constraint systems as "very powerful because they can represent any efficiently 
verifiable program" [[1]]. 

### R1CS Factorization Example for Bulletproofs 

> In the article [[16]], Harchandani explores the Dalek Bulletproofs API with various examples. Of interest

Reviewer: ```suggestion
In [[16]], Harchandani explores the Dalek Bulletproofs API with various examples. Of interest
```

is the factorization problem, which is one out of the six R1CS Bulletproof examples 
discussed in the article. 
The computational challenge is to "_prove knowledge of factors_ p 
_and_ q _of a given number_ r _without revealing the factors_". 

Table 3 gives an outline of the description and the code lines of the example. Harchandani's complete 

> code of this example can be found in [[17]]. 

Reviewer: ```suggestion
code of this example can be found in [[17]]. Important to note is that the verifier must have an exact copy of the R1CS 
circuit used by the prover.
```

"Constraint Systems for ZK SNARKs" 

[[8]] H. Wu, W. Zheng, A. Chiesa, R. Ada Popa, and I. Stoica, "DIZK: A Distributed Zero Knowledge 
Proof System", Proceedings of the 27th USENIX Security Symposium, August 15–17, 2018 [online]. Available: 
<https://www.usenix.org/system/files/conference/usenixsecurity18/sec18-wu.pdf>. Date accessed: 2019&#8209;12&#8209;14. 

[8]: https://www.usenix.org/system/files/conference/usenixsecurity18/sec18-wu.pdf
"DIZK: A Distributed Zero 
Knowledge Proof System" 

[[9]] E. Ben-Sasson, A. Chiesa, D. Genkin, E. Tromer and M. Virza, "SNARKs for C: Verifying Program 
Executions Succinctly and in Zero Knowledge (extended version)," October 2013 [online]. Available: 
<https://eprint.iacr.org/2013/507.pdf>. Date accessed: 2019&#8209;12&#8209;17. 

[9]: https://eprint.iacr.org/2013/507.pdf
"SNARKs for C: 
Verifying Program Executions 
Succinctly and in Zero 
Knowledge (extended version)"

[[10]] V. Buterin, "Quadratic Arithmetic Programs: from Zero to Hero," 12 December 2016 [online]. Available: 
<https://medium.com/@VitalikButerin/quadratic-arithmetic-programs-from-zero-to-hero-f6d558cea649>. Date accessed: 2019&#8209;12&#8209;19.

> [10]: https://medium.com/@VitalikButerin/quadratic-arithmetic-programs-from-zero-to-hero-f6d558cea649	"Quadratic Arithmetic Programs: from Zero to Hero"

Reviewer: ```suggestion
[10]: https://medium.com/@VitalikButerin/quadratic-arithmetic-programs-from-zero-to-hero-f6d558cea649 "Quadratic Arithmetic Programs: from Zero to Hero"
```

Reviewer: Another Gremlin that broke the link

"Ristretto"

[[14]] H. Valence, "Bulletproofs pre-release", [online]. Available: <https://medium.com/interstellar/bulletproofs-pre-release-fcb1feb36d4b> Date accessed: 2019&#8209;11&#8209;21.

[14]: https://medium.com/interstellar/bulletproofs-pre-release-fcb1feb36d4b
"Bulletproofs pre-release"

[[15]] Dalek, "Bulletproofs Implementation", [online]. Available:
<http://github.com/dalek-cryptography/bulletproofs/> Date accessed: 2019&#8209;10&#8209;02.

[15]: http://github.com/dalek-cryptography/bulletproofs/
"Dalek's Bulletproofs Implementation"

[[16]] L. Harchandani, "Zero Knowledge Proofs using Bulletproofs" [online]. Available: 
<https://medium.com/coinmonks/zero-knowledge-proofs-using-bulletproofs-4a8e2579fc82>. Date accessed: 2020&#8209;01&#8209;03.

[16]: https://medium.com/coinmonks/zero-knowledge-proofs-using-bulletproofs-4a8e2579fc82
"Zero Knowledge Proofs 
using Bulletproofs"

[[17]] L. Harchandani, "Factors R1CS Bulletproofs Example", [online]. Available: 

<https://github.com/lovesh/bulletproofs/blob/e477511a20bdb8de8f4fa82cb789ba71cc66afd8/tests/basic_r1cs.rs#L17> Date accessed: 2019&#8209;10&#8209;02. 

> [17]: https://github.com/lovesh/bulletproofs/blob/e477511a20bdb8de8f4fa82cb789ba71cc66afd8/tests/basic_r1cs.rs#L17	"Factors R1CS Bulletproofs Example"

Reviewer: ```suggestion
[17]: https://github.com/lovesh/bulletproofs/blob/e477511a20bdb8de8f4fa82cb789ba71cc66afd8/tests/basic_r1cs.rs#L17 "Factors R1CS Bulletproofs Example"
```

Reviewer: There was a Gremlin hidden in there, which broke the link...

##### 

> # **Rank-1 Constraint System with Application to Bulletproofs**

Author: Done

[[1]] suggests this is true. One of the Interstellar team's goals is to use the constraint system in their 
_Confidential Asset Protocol_ called the *Cloak* and in their envisaged [_Spacesuit_](https://github.com/interstellar/spacesuit). 
Despite their work on using R1CS being research in progress, their detailed notes on constraint systems and their 
implementation in RUST are available in [[13]]. 

Most recent constructions of zk-SNARKs and Bulletproofs involve _arithmetic circuits_ and _R1CSs_, 
either explicitly or implicitly. 

The aim of this report is to: 

- highlight the connection between arithmetic circuits and R1CSs; 
- clarify the difference R1CSs make in Bulletproofs and in range proofs; 
- compare R1CSs in zk-SNARKs and Bulletproofs; 
- explain what is meant by the following in [[12]] - "We note that a range proof 
  using the protocol of [[BCC+16]](Bootle et al's paper [[4]]) would have required implementing the commitment opening 
  algorithm as part of the verification circuit, which we are able to eliminate." 

## Arithmetic Circuits

### Overview 

Arithmetic circuits are said to be the most natural and standard model for computing polynomials. Every 

> function $ \mathcal{H} : \lbrace 0 , 1 \rbrace^n → \lbrace 0 , 1 \rbrace^m $ of fixed input and output _length_ 

Author: Done

can be represented as an arithmetic circuit over any finite field $ \mathcal{F}_p$ [[2]]. 

> According to [[3]], arithmetic circuits are a more highly structured model of computation 

Author: Done

than Boolean circuits. For instance, when studying arithmetic circuits, one is interested in _syntactic_ 
computation of polynomials, whereas in studying Boolean circuits, one is interested in the _semantics_ of the computation. 
In the Boolean case, one is not interested in any specific polynomial representation of the function, 
but only wants to compute some representation of the function. In the arithmetic case, one focuses on a 
_specific representation_ of the function. 

> For our purposes, arithmetic circuits have as inputs variables $ x_1, \dots , x_n $. Computations are 

Author: Done

performed using the arithmetic operations $ + $ and $ × $, and may involve constants from a field $\mathcal{F}_p$. 
Complexity measures associated with such circuits are _size_ and _depth_, which capture _the number of gates_ in a 

> circuit and _the maximum distance between an input and an output_, respectively [[3]].

Author: Done

  it has $k$ outgoing wires. 
- Every gate in $\mathcal{A}$ of _in-degree_ 0 is labeled by either a variable ${x_i}$ or some field 
  element from $\mathcal{F}$. Such a gate is called an _input gate_. Every gate of _out-degree_ 0 is called an 
  _output gate_. 
- Every other gate in $\mathcal{A}$ is labeled by either $\otimes$ or $\oplus$, and called a 
  _multiplication gate_ or _addition gate_, respectively. We are interested in _binary_ circuits, i.e. circuits where 
  each gate has two _input wires_ and one _output wire_. 
- An _arithmetic circuit_ is called a _formula_ if it is a _directed tree_ whose edges are _directed_ from the leaves 
  to the root. 
- The size of $\mathcal{A}$, denoted $|\mathcal{A}|$, is the number of _wires_, i.e. _edges_, in $\mathcal{A}$. 

A typical multiplication gate has a _left input_ $a_L$, a _right input_ $a_R$ and 
an _output_ $a_O$. Also, we note that $ a_L \cdot a_R - a_O = 0 $.

<div align="center"><b>
<img src="sources/basic-multiplication-gate.png" alt="basic-multiplication-gate" style="zoom:67%;" />
</b></div> 
<div align="center"><b>Figure 1: Typical Multiplication Gate</b></div> 

> We note that in cases where the inputs and outputs are all _vectors_ of 

Author: Done

$n$ components, i.e. $\mathbf{a_L} = ( a_{L, 1}, a_{L, 2} , \dots , a_{L, n})$, 
$\mathbf{a_R} = ( a_{R, 1}, a_{R, 2} , \dots , a_{R, n})$ 
and $\mathbf{a_O} = ( a_{O, 1}, a_{O, 2} , \dots , a_{O, n})$, 
then _multiplication_ of $\mathbf{a_L}$ and $\mathbf{a_R}$ is defined as an _entry-wise_ 
product called the _**Hadamard product**_; 

$$
\mathbf{a_L}\circ \mathbf{a_R} = \big(( a_{L, 1} \cdot a_{R, 1} ) , ( a_{L, 2} \cdot a_{R, 2} ) , \dots , ( a_{L, n} \cdot 
a_{R, n} ) \big) =  \mathbf{a_O}
$$

The equation ${ \mathbf{a_L}\circ \mathbf{a_R} = \mathbf{a_O} }$ is referred to as a _multiplicative constraint_, 
but is also known as the _Hadamard relation_ [[4]]. 

### Example of Arithmetic Circuit 

An arithmetic circuit computes a polynomial in a natural way, as shown in this example. Consider the following 
arithmetic circuit $\mathcal{A}$ with inputs $\lbrace x_1 , x_2 , 1 \rbrace$ over some field $\mathcal{F}$: 

<div align="center"><b>
<img src="sources/polynomial-eg-ac.png" alt="polynomial-eg-ac" style="zoom:67%;" />
</b></div> 
<div align="center"><b>Figure 2: Arithmetic Circuit</b></div> 

> The output of $\mathcal{A}$ above is the polynomial $x^2_1 \cdot x_2 + x_1 + 1 $ of _total degree_ three. 

Author: Done

than Boolean circuits, it also points out that "we do not know how to efficiently reconstruct a circuit using only queries to 
the polynomial it computes". 

Zero-knowledge (ZK) proofs, such as zk-SNARKs and Bulletproofs, require that _statements to be proved_ are expressed in their 
simplest terms for efficiency. According to [[5]], "the ZK SNARK end-to-end journey is to create a _function_ 
to write proofs about", but they "must work with specific constructs". That is why arithmetic circuits are so important 
in making ZK more efficient: "these _functions_ have to be specified as sequences of very simple terms, 
namely, additions and multiplications of only two terms in a particular field" [[5]].

In verifying a zk-SNARK proof, the verifier needs to carry out a step-by-step check of the computations, i.e. for 
each gate, the verifier has to check if the _output_ $ a_O $ is correct with respect to the given _inputs_ 
$ a_L $ and $ a_R $, and test if $ a_L \cdot a_R - a_O = 0 $ for each multiplication 
gate. This requires that an _addition gate_ be treated as some form of a _multiplication gate_. This is explained later in 
this report. 

## Rank-1 Constraint Systems 

> ### Overview

Author: Done

Although a computation problem is typically expressed in terms of a high-level programming language, a zk-SNARK requires 
it to be expressed in terms of a set of _quadratic constraints_, which is closely related to circuits of logical gates. 

> R1CS is short for Rank-1 Constraint System. These types of constraint systems have been featured in several 

Author: Done

constructions of zk-SNARKs. At times they were simply referred to as _quadratic constraints_ or _quadratic equations_; refer to 
[[6]], [[7]] and [[8]]. 

> ### Definition of Constraint System 

Author: Done

A constraint system was originally defined by Bootle 
et al in [[4]]. The Dalek team give a more general definition of a constraint system in [[10]]: 

"A **constraint system** is a collection of arithmetic constraints over a set of variables. There are two kinds of 
variables in the constraint system:

1. ${m}$ **high-level** variables, the input secrets ${ \mathbf{v}}$, 
2. $ n$ **low-level** variables, the internal input vectors ${ \mathbf{a}_L}$ and 
   ${ \mathbf{a}_R},$ and output vectors ${ \mathbf{a}_O } $ of the multiplication gates." 

> Specifically, a **Rank-1 Constraint System** **(R1CS)** is a system that consists of two sets of constraints: 

Author: Done

- ${ n}$ _multiplicative constraints_, $ \mathbf{ a_L \circ a_R = a_O } $, where "${ \circ }$" is the Hadamard product; and

> - ${ q}$ _linear constraints_, $\mathbf{W_L\cdot { a_L} + W_R\cdot { a_R} + W_O\cdot { a_O } = W_V\cdot { v + c} } $, where $\mathbf{W_L, W_R, W_O}$ and $\mathbf{W_V}$ are weights applied to respective input vectors and output vectors [[10]]. 

Author: Done

Note that it was Bootle et al. who first expressed _arithmetic circuit satisfiability_ in terms of the _Hadamard relation_ 
and linear constraints [[10]]. In their definition, the above linear constraints are written as:

$$
\mathbf{W_L\cdot { a_L} + W_R\cdot { a_R} + W_O\cdot { a_O } = c }
$$

> That is, without the vector $\mathbf{v}$ and its weight $\mathbf{W_V} $. This is explained later. 

Author: Done

> ### Definition of Rank-1 Constraint System 

Author: Done

This paragraph provides a simplified definition of an R1CS as it applies to zk-SNARKs [[9]]. 

An R1CS is a sequence of groups of three vectors ${ \bf{a_L}}, { \bf{a_R}}, { \bf{a_O}} ,$ and the 
solution to an R1CS is a vector ${ \bf{s}}$ that satisfies the equation:

$$

> { \langle {\bf{a_L , s}} \rangle \cdot \langle {\bf{a_R , s }} \rangle - \langle {\bf{a_O , s }} \rangle = 0 }

Author: Done

| 1.   | Introduces the Hadamard relation and linear constraints.     | Turns the Hadamard relation and linear constraints into a single linear constraint, and these are in fact the R1CS. | Generalizes constraint systems and uses what is called gadgets as building blocks for constraint systems. |
| 2.   | Improves on Groth's work [[15]] on ZK proofs. Reducing a $\sqrt{N}$  complexity to  $6log_2(N) + 13$, where $N$  is the circuit size. | Improves on Bootle et al.'s work [[4]]. Reducing a $2log_2(N) + 13$ complexity to  $6log_2(N) + 13​$, where $N​$ is the circuit size. | Adds constraint systems to Bunz et al.'s work on Bulletproofs, which are short proofs, and the complexity advantage is seen in proving several statements at once. |
| 3.   | Introduces logarithm-sized inner-product ZK proofs.          | Introduces Bulletproofs, extending proofs to proofs of arbitrary statements. The halving method is used on the inner-products, resulting in the above reduction in complexity. | Introduces gadgets that are actually add-ons to an ordinary ZK proof. A range proof is an example of a gadget. |

> | 4.   | Uses Fiat-Shamir heuristics in order to achieve non-interactive ZK proofs. | Bulletproofs have no trusted setup, as they use  the Fiat Shamir heuristics to achieve non-interactive ZK proofs. | Merlin transcripts are specifically used for a Fiat-Shamir transformation to achieve non-interactive ZK proofs. |

Author: Corrected

| 5.   | The Pedersen commitments are used in order to achieve ZK property. | Eliminates the need for a commitment algorithm by including Pedersen commitments among the inputs to the verification proof. | Low-level variables, representing inputs and outputs to multiplication gates, are computed per proof and committed using a single vector Pedersen commitment. |
| 6.   | The ZK proof involves conversion of the arithmetic circuit into an R1CS. | The mathematical expression of a Hadamard relation is closely related to an arithmetic circuit. The use of this relation plus linear constraints as a single constraint amounts to using a constraint system. | Although arithmetic circuits are not explicitly used here, the Hadamard relation remains the same as first seen in Bulletproofs, more so in the inner-product proof. |

> | 7.   | The ZK proof here is for NP statements based on the difficulty of the discrete logarithm problem. | As mentioned above, Bulletproofs extend ZK proofs (such as range proofs) to proofs on arbitrary statements. They are also described as short non-interactive proofs for arithmetic circuits without a trusted setup. | Interstellar is building an Application Programming Interface (API) that allows developers to choose their own collection of gadgets suitable for the protocol they wish to develop. |

Author: NP defined. 
The significance of the row is in indicating the improvements made that led to achieving programmable constraint systems, ... the programmable constraint system is robust in the sense that a developer can choose their collection of gadgets.

## zk-SNARK Constraint Systems 

> ### Overview 

Author: Sorted.

How exactly does one go _**from an arithmetic circuit to an R1CS**_ ? 

Eli Ben-Sasson et al. [[8]] mention that the difference in computational power among the parties makes _succinct verification_ 
a requirement. Hence, the pursuit of short and fast proofs systems. These make it possible for a verifier to check 
a non-deterministic polynomial-time computation in a much shorter time than the time required to run the computation 
when given a valid NP witness. They also state that the difficulty in studying the efficiency of proofs for NP 
statements is the _problem of representation_. Arithmetic circuits and the R1CS are some of such _representations_ of 
statements to be proved. 

> ### Summary 

Author: No more summaries in the report.

In his blogpost entitled "_Constraint Systems for ZK SNARKs_" [[5]], 

> Alex Pinto wrote with the intention to elucidate the difficulties he came across while making "DIZK, a library written 

Author: Removed from the report.

> in JAVA, work with the same zkSNARK scheme as used by ZoKrates and libsnark, which are written in C++". He says the 

Author: zkSNARKs removed from the report.

> biggest obstacle was "proper encoding of the function that we want to prove statements for". So he focuses specifically 

Author: Removed from the report.

circuit into an R1CS_. Although Pinto covers the same concepts covered by Buterin in [[9]] (there is a summary of 
Buterin's article in [zk-SNARKs](https://tlu.tarilabs.com/cryptography/zksnarks/mainreport.html#quadratic-arithmetic-programs---from-zero-to-hero)), 
Pinto gives some finer details. 

- The _first step_ is to express the given computational problem in terms of _addition_ and _multiplication_ of only two 
  terms. This involves introducing _auxiliary variables_ in order to express the given computational problem in terms of 
  _equations_ with a single operation between two variables. The result is an arithmetic circuit.
- The _second step_ is to encode the arithmetic circuit as an R1CS.

Pinto explains how this is achieved. He makes it easy to understand how the _solution vector_ as shown in Table 1 is 
constructed, as well as how addition gates need to be expressed as some _multiplication gate_ in order to understand the 
resulting R1CS. Pinto uses an example inspired by RSA's well-known _difficulty of factorization problem_: "Proving that 
one knows prime factors $ p $ and $ q $  such that $ n + 1 = (p + 3)(q + 2) $." 

> When comparing zk-SNARKs and Bulletproofs, Lovesh Harchandani in [[14]] mentions the two main 

Author: Sorted.

disadvantages of zkSNARKs: 

- Firstly, there is a _trusted setup_ used for a one-time generation of protocol parameters (e.g. a so-called 
_Common Reference String_ (CRS)). The problem with such a setup is that the generated parameters (or the CRS) could be 
used to leak some secrets. One way of avoiding such possible leakage of secrets is to use a Multiparty 
Computation (MPC) as utilized in cryptocurrencies such as Zcash. 
- Secondly, a new _trusted setup_ (i.e. a fresh CRS) is needed for each circuit, 
simply because each new computational challenge will have different constraints. 

The Bulletproofs constraint system, on the other hand, has no trusted setup. 

## Interstellar's Bulletproof Constraint System

> ### Overview 

Author: Done. thanks

> ### Overview 

Author: Done

of a prime-order group using a cofactor-8 curve known as Curve25519. They reported on how they implemented Bulletproofs 
in Henry de Valence's article entitled "[Bulletproofs pre-release](https://medium.com/interstellar/bulletproofs-pre-release-fcb1feb36d4b)". 
An update on their progress in extending the [Bulletproofs implementation](http://github.com/dalek-cryptography/bulletproofs/) 
to a constraint system API, which enables ZK proofs of arbitrary statements, was given in Cathie Yun's 
article, [Programmable Constraint Systems for Bulletproofs](https://medium.com/interstellar/programmable-constraint-systems-for-bulletproofs-365b9feb92f7) 
[[11]]. However, it was Bootle et al. [[4]] who first used the Hadamard relation and linear constraints, which together 
form the constraint system as formalized by the Interstellar team. Most of the mathematical background of these 
constraints and bulletproofs is contained Bunz et al.'s paper [[12]]. For an extensive study on Bulletproof 
protocols, refer to the previous report [here](https://tlu.tarilabs.com/preface/learning/bulletproofs.html). 

> Dalek's constraint system, as defined earlier in [Definition of Constraint System](#definition-of-constraint-system), is a collection of arithmetic constraints of two types, 

Author: Not clear what to change. Link looks fine.

4. The verifier now _verifies the proof_ by enforcing the _same constraints_ over _the commitments_ plus any public 
   values.

### About Gadgets 

Consider a _verifiable shuffle_: given two lists of _committed values_ ${ x_1, x_2, . . . , x_n}$ and 
${ y_1, y_2, . . . , y_n} ,$ prove that the second list is a permutation of the first. Bunz et al. (page 5 of [[9]]) mention that 
the use of bulletproofs improves the complexity of such a _verifiable shuffle_ to size $\mathcal{O}(log(n))$ compared to 
previous implementation results. Although not referred to as a _gadget_ in the paper, this is in 
fact a _shuffle gadget_. The term _gadget_ was used and popularized by the Interstellar team, who introduced _gadgets_ 
as building blocks of constraint systems; refer to [[1]]. 

A _shuffle gadget_ is any function whose outputs are but a permutation of its inputs. By definition of a permutation, the 
_number of inputs_ to a shuffle gadget is always the same as the _number of outputs_. 

<div align="center"><b> 
<img src="sources/simple-shuffle-diagram.png" style="zoom:67%;" /> 
</b></div> 
​				

<div align="center"><b>Figure 3: Simple Shuffle Gadgets with Two Inputs [[1]]</b></div> 

> Find a RUST example code for a _shuffle gadget_ in the _Bulletproofs Constraint System_ framework 

Author: Done

[here](https://github.com/lovesh/bulletproofs/blob/e477511a20bdb8de8f4fa82cb789ba71cc66afd8/docs/r1cs-docs-example.md), 

> written by Lovesh Harchandani. 

Author: Done

## Conclusions, Observations and Recommendations 

In conclusion, this report has: 

- highlighted the connection between arithmetic circuits and R1CSs; 
- clarified the difference R1CSs make in Bulletproofs and in range proofs; 
- compared R1CSs in zk-SNARKs and Bulletproofs; 
- explained what is meant by the following in [[12]] - "We note that a range proof 
  using the protocol of [[BCC+16]](Bootle et al.'s paper [[4]]) would have required implementing the commitment opening 
  algorithm as part of the verification circuit, which we are able to eliminate." 

Constraint systems indeed form a natural language for most computational problems expressed as arithmetic circuits. No 
wonder they have found such ample application in both zk-SNARKs and Bulletproofs. 
Although much work still needs to be done, Bulletproofs with constraint systems built on them promise to be powerful 
tools for efficient handling of verifiable programs. The leverage that developers have, in choosing whatever gadgets 
they wish to implement, leaves enough room to build proof systems that have some degree of modularity. 

> The possibility of using this Bulletproofs framework in building confidential digital asset schemes provides 

Author: Done

> greater opportunities or options to consider when building Tari's envisaged Digital Assets Network. 

Author: Made a comment on the conclusion.

> greater opportunities or options to consider when building Tari's envisaged Digital Assets Network. 

Author: Sorted.

Bulletproofs. Only time will tell which of these two will yield better practicality. 

## References 

[[1]] C. Yun, "Building on Bulletproofs" [online]. Available: <https://medium.com/@cathieyun/building-on-bulletproofs-2faa58af0ba8>. 
Date accessed: 2020&#8209;01&#8209;03.

[1]: https://medium.com/@cathieyun/building-on-bulletproofs-2faa58af0ba8
"Building on Bulletproofs"

[[2]] A. Szepieniec and B. Preneel, "Generic Zero-knowledge and Multivariate Quadratic Systems", 2016 [online]. 
Available: <https://pdfs.semanticscholar.org/06c8/ea507b2c4aaf7b421bd0c93e6145e3ff7517.pdf?_ga=2.124585865.240482160.1578465071-151955209.1571053591>. 
Date accessed: 2020&#8209;01&#8209;08. 

[2]: https://pdfs.semanticscholar.org/06c8/ea507b2c4aaf7b421bd0c93e6145e3ff7517.pdf?_ga=2.124585865.240482160.1578465071-151955209.1571053591
"Generic Zero-knowledge and 
Multivariate Quadratic Systems"

[[3]] A. Shpilka and A. Yehudayoff, "Arithmetic Circuits: A Survey of Recent Results and Open Questions", 
Technion-Israel Institute of Technology, Haifa, Israel, 2010 [online]. Available: 

> <https://ieeexplore.ieee.org/document/8186881/metrics#metrics>. 

Author: Done

Date accessed: 2019&#8209;12&#8209;21. 

> [3]: https://ieeexplore.ieee.org/document/8186881/metrics#metrics

Author: Done
